### DISCUSSION TITLE
Tabs for opened requests

### DISCUSSION THREAD

**Discussion by NickRimmer** [Kong/insomnia#6108] [Show and Tell]
Hey there!

First off, I just want to say that Insomnia is such an amazing tool! It's a refreshing alternative to Postman, packed with its own unique and cool features - kudos to the team behind it! 🎉

However, as a user, I found myself missing tabs for 'opened' requests. Sure, there's a handy quick switch via hotkeys, but in my workflow, I tend to be more of a mouse-clicker, jumping around in the UI between requests. So, I thought - why not bring tabs to the party?

I've brewed up a plugin to do just this! Meet "Tabs: Requests Navigator". It's all set to make your Insomnia experience even more seamless. Install it directly by its name `insomnia-plugin-request-navigator` or grab it from the plugin page.
https://insomnia.rest/plugins/insomnia-plugin-request-navigator

<img src="https://github.com/NickRimmer/insomnia-plugin-request-navigator/raw/main/npm/content/demo-2.jpg"/>

I hope it adds a sprinkle of convenience and fun to your Insomnia journey. Cheers! 🚀

**Comment by smiller0924** [comment]
I cannot get the tabs to work.  I have the following version of Insomnia:

Insomnia 2023.5.8
Build date: 9/7/2023
OS: Windows_NT x64 10.0.19045
Electron: 25.2.0
Node: 18.15.0
Node ABI: 116
V8: 11.4.183.23-electron.0
Architecture: x64

Is there a hotkey I have to click to enable this?  Thank you.

**Comment by tarciomota** [comment]
no longer working
```
Error: Cannot find module 'nedb'
Require stack:
- C:\Users\<USER>\AppData\Local\insomnia\app-8.6.1\resources\app.asar\index.html
    at Module._resolveFilename (node:internal/modules/cjs/loader:1084:15)
    at o._resolveFilename (node:electron/js2c/renderer_init:2:3225)
    at Module._load (node:internal/modules/cjs/loader:929:27)
    at l._load (node:electron/js2c/asar_bundle:2:13642)
    at o._load (node:electron/js2c/renderer_init:2:2455)
    at Module.require (node:internal/modules/cjs/loader:1150:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (C:\Users\<USER>\AppData\Roaming\Insomnia\plugins\insomnia-plugin-request-navigator\index.js:3202:19)
    at Object.<anonymous> (C:\Users\<USER>\AppData\Roaming\Insomnia\plugins\insomnia-plugin-request-navigator\index.js:3774:3)
    at Module._compile (node:internal/modules/cjs/loader:1271:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1326:10)
    at Module.load (node:internal/modules/cjs/loader:1126:32)
    at Module._load (node:internal/modules/cjs/loader:967:12)
    at l._load (node:electron/js2c/asar_bundle:2:13642)
    at o._load (node:electron/js2c/renderer_init:2:2455)
    at Module.require (node:internal/modules/cjs/loader:1150:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at cG (file:///C:/Users/<USER>/AppData/Local/insomnia/app-8.6.1/resources/app.asar/index-db9674c5.js:81:7484)
    at kk (file:///C:/Users/<USER>/AppData/Local/insomnia/app-8.6.1/resources/app.asar/index-db9674c5.js:81:8221)
    at async u (file:///C:/Users/<USER>/AppData/Local/insomnia/app-8.6.1/resources/app.asar/index-db9674c5.js:1512:17437)
```

**Reply by Deep-Adhikary** [reply]
@tarciomota 
You need to change the nedb import as insomania is using nedb fork
1. Go to plugin folder
2. Open index.js. 
3. Then change: "var NeDb = window.require("nedb"); " to "var NeDb = window.require("@seald-io/nedb");"

Please refer to https://github.com/Kong/insomnia/issues/7019 for more details.
It would be good if Insomania adopt this. as @NickRimmer no longer maintaining the plugin.

**Reply by NickRimmer** [reply]
Yeah, sorry for this. I simply cannot test changes anymore, as I do not have latest insomnia installed -_-

**Reply by venkateshgopalakrishnan** [reply]
@Deep-Adhikary can't thank you enough for posting this fix. I should have come across this answer long back.

**Comment by subnetmarco** [comment]
We will start work to officially support tabs within the next 3 months, maybe sooner.

**Reply by smiller0924** [reply]
I just wanted to see if you are adding the tabs feature soon in Insomnia?  Thanks.

**Reply by subnetmarco** [reply]
Yes! We prioritized the scripting capabilities (pre-request scripting in 9.0 and [after-response scripting](https://github.com/Kong/insomnia/discussions/6242#discussioncomment-9533769) in the upcoming 9.3). We are also now working on implementing [globals environments](https://github.com/Kong/insomnia/discussions/3694#discussioncomment-9593245), and after we will like to add tabs soon.

It will look like this:

![Screenshot 2024-05-29 at 11 03 02 PM](https://github.com/Kong/insomnia/assets/964813/e778be82-a4ac-4e36-892c-81e67b2dacc3)

**Reply by smiller0924** [reply]
I think that looks pretty cool!  

Will you also be adding a Visualizer feature in the future? This way we can take relies and display the information in an custom HTML window.   That would be next level.

Keep up the good work!

**Reply by subnetmarco** [reply]
> Will you also be adding a Visualizer feature in the future?

That's the plan. What visualizations types do you like to display?

**Reply by smiller0924** [reply]
Sorry I didn't get back to you.  Typically I just build very simple HTML pages to display elements of my REST API Replies.  Here is an example:

`let vizTemp = `
<style type="text/css">
table {
font-size:12px;
color: gray;
width:85%;
border-width: 1px;
border-color: "black";
border-collapse: separate;
}
table.center {
  margin-left: auto; 
  margin-right: auto;
}

        td {
            padding: 20px
        }

    </style>

    <div class="header">
    <h1>Volume Quote</h1>
    {{#each response.statusList}}
        <p>{{description}}</p>
    {{/each}}
    <p>Quote ID: {{response.quoteId}}</p>
    <p>Effective Date: {{response.effectiveDate}}</p>
    <p>Expiration Date: {{response.expirationDate}}</p>
    </div>   

    <table border="1">
    <table class="center">
    <tr>
        <th style="background-color: blue">Service</th>
        <th style="background-color: blue">Delivery</th>
        <th style="background-color: blue">Rate (USD)</th>       
    </tr>

    {{#each response.options}}
        <tr>
            <td>{{optionType}}</td>
            <td>{{serviceLevel}}</td>
            <td>{{rate}}</td>
        </tr>
    {{/each}}

</table>`

    pm.visualizer.set(vizTemp, {
        response: pm.response.json()
    });
`
![image](https://github.com/Kong/insomnia/assets/108164566/ab50523d-c919-4e87-b8ba-3b64c92f714c)

Here is another example:

`var jsonData = JSON.parse(responseBody);

pm.environment.set("serviceType", jsonData.output.rateReplyDetails[0].serviceType);
pm.environment.set("packageType", jsonData.output.rateReplyDetails[0].packagingType);

let vizTemp = `
<style type="text/css">
table {
font-size:12px;
color: gray;
width:85%;
border-width: 1px;
border-color: "black";
border-collapse: separate;
}
table.center {
  margin-left: auto; 
  margin-right: auto;
}

        td {
            padding: 20px
        }

    </style>   

    <table border="1">
    <table class="center">
    <tr>
        <th style="background-color: blue">Service</th>
        <th style="background-color: blue">Delivery</th>
        <th style="background-color: blue">List (USD)</th>
        <th style="background-color: blue">Net (USD)</th>
        <th style="background-color: blue">List (OTHER)</th>
        <th style="background-color: blue">Net (OTHER)</th>        
    </tr>

    {{#each response.output.rateReplyDetails}}
        <tr>
            <td>{{serviceName}}</td>
            <td>{{commit.dateDetail.dayFormat}}</td>
            {{#each ratedShipmentDetails}}
                <td>{{totalNetFedExCharge}}</td>
            {{/each}}
        </tr>
    {{/each}}

</table>`

pm.visualizer.set(vizTemp, {
response: pm.response.json()
});`

![image](https://github.com/Kong/insomnia/assets/108164566/dd8ea0f4-d11b-4987-a6b2-0d5e2e0dc93e)

I hope that this is helpful.

**Reply by smiller0924** [reply]
Hi there!  I just wanted to check in with you in the Vizualization feature I asked about.  Is there a feature in Insomnia similar to the one in Postman?  Were my examples helpful?  Please let me know your thoughts. Thank you.

**Comment by CurryYangxx** [comment]
Hi there!  The officially support "Multi-tabs" feature has been finished development, you can download the artifact [here](https://github.com/Kong/insomnia/actions/runs/12002861045?pr=8151) and try it out. We’d appreciate your suggestions and feedback to help improve it further. 
We plan to ship it in v11 and hope this feature can help you enhance your efficiency. Thanks!

PR: https://github.com/Kong/insomnia/pull/8151


https://github.com/user-attachments/assets/ee2fd9f2-db7b-40ae-8837-68a2d62e5aed

**Reply by lmorganl** [reply]
seems to be nice on W11 with a large collection.
Maybe will be better to open only request on tab. Open folder on tab will not be usefull (only my point of view)

**Reply by CurryYangxx** [reply]
Current Insomnia support configures some shared stuff at the folder level, such as headers, scripts, env, etc, so I think it will also be helpful.

**Reply by lmorganl** [reply]
Is it possible to see how many tabs are open?
If too many tabs are open, they are not displayed.

**Reply by venkateshgopalakrishnan** [reply]
> Is it possible to see how many tabs are open? If too many tabs are open, they are not displayed.

I see that we can scroll to see all the open tabs. But finding the one amongst all the open tabs is sometimes a little difficult. A button that will hide the extra tabs and show the number of tabs opened in addition to what can be held by the tabs bar might help. 
But in my opinion, this is fine too.

**Reply by venkateshgopalakrishnan** [reply]
Also rearranging of tabs isn't supported. That would be a beneficial feature.

**Reply by CurryYangxx** [reply]
https://github.com/user-attachments/assets/08390af9-c41a-4b8e-ab0a-d9695b8d35bb

@venkateshgopalakrishnan Good feedback! 
We made some improvements recently, as you can see in this screen recording

- auto scroll into view whenever you click a tab or a request on the left side
- support drag and drop to re-arrange tabs

download the latest artifact to experience: https://github.com/Kong/insomnia/actions/runs/12134727321?pr=8151

**Reply by lmorganl** [reply]
Win install doesn't work for me (i'm will waiting beta release)

**Reply by CurryYangxx** [reply]
@lmorganl Do you have any problems with Windows?

**Reply by lmorganl** [reply]
> @lmorganl Do you have any problems with Windows?

There's an error installing the version 10.2.1-8f87589.pr-8151.0.exe. That's all ;)

**Reply by Deep-Adhikary** [reply]
This is indeed a great usability improvement. Excieted to see it in main brunch.

**I feel pinned tab(s) will be a good scope fore future improvement(if this is too complex to be implemented)** 

I did not used insomania around 6+ months. It indeed comes long way and the app looks great. I have seen `scripting` is also implemented(Thanks to this, I could ditch postman :D)

**Reply by lmorganl** [reply]
Do you have an idea of when tabs will be release ?

**Reply by subnetmarco** [reply]
It will ship in January/February in Insomnia v11.

**Comment by CurryYangxx** [comment]
Hey everyone! The multiple-tabs feature has been released in the 11.0.0-beta.0 version, including these features:

- Create tab automatically
- Could between multiple requests in different workspaces or projects under the same org.
- Support drag and drop to re-arrange
- Close all tabs&other tabs by right-clicking
- ShortCut for adding a new request tab

You can download it here: https://github.com/Kong/insomnia/releases/tag/core%4011.0.0-beta.0.
Please give it a try, and looking forward to hearing your feedback!

**Comment by lmorganl** [comment]
Hello

my first feedback :
- Add an option not to open a new tab when clicking on a  folders (i have to many folders and need to close tabs... boring)
- in scratch pad, when close the last tab, we quit mode.

**Reply by subnetmarco** [reply]
> Add an option not to open a new tab when clicking on a folders (i have to many folders and need to close tabs... boring)

@CurryYangxx

**Reply by CurryYangxx** [reply]
Fix the first one in this issue: https://github.com/Kong/insomnia/issues/8433
And the second one will discuss with the team how to improve it.

**Reply by CurryYangxx** [reply]
Hi! We will support temporary tabs to improve your experience.
- https://github.com/Kong/insomnia/pull/8486

This is the demo:

https://github.com/user-attachments/assets/a6f5b8c6-a310-4b06-bf28-7195003a7564

**Comment by victorlapshev** [comment]
Is there is an option to disable tabs?

**Reply by subnetmarco** [reply]
@victorlapshev Do you think that introducing support for temporary tabs will alleviate some of the issue you may have with the functionality: https://github.com/Kong/insomnia/discussions/6108#discussioncomment-12549245

**Reply by victorlapshev** [reply]
I don’t think so — I just don’t like the tabbed interface in the REST client. But that’s just my personal taste; others might feel differently

**Comment by krystofremes** [comment]
Hi, just wanna add my 2 cents here..

Tbh current implementation of the tabs feature makes me wanna tear all my hair out. The core idea is great. However opening a tab for every single left click by default is imho a mistake. Looked into the temporary tab [PR](https://github.com/Kong/insomnia/pull/8486) and it seems unnecesarily complicated.

As the app feels like a browser and pretty much is a browser, wouldnt it make sense to implement behaviour typical for a browser? I find myself intuitively reaching for a middle click for opening something as a new tab and similarly for closing a tab. Sticking to a "left click opens in this tab" and "middle click opens in a new tab".

As I typed this I kinda realized, it might collide with browser behaviour for users using insomnia in a actual web browser, though it still feels as a more intuitive behaviour.

EDIT: Adding some context: 
My first interaction with the tabs feature happened today as I imported a vendors openapi spec into insomnia to look through. And lo and behold every single endpoint I click on opens in a new tab resulting in a million of tabs opened, which is quite a frustrating experience.

**Comment by Eduardishion** [comment]
Hello, my insomnia was updated to version 11.0.0, and in particular the use of tabs I do not see very useful, I reach a point where I have many open taps, I wanted to know if there is any way to deactivate them.


Consider that it loses its aesthetic of simplicity and saturates the content of the window by having so many tabs open

**Comment by vplagov** [comment]
I would also like an option to disable tabs altogether. I'm not used to tabs. I always have the left sidebar open with all my requests and folders. And I'm so used to navigating through my requests within that sidebar. If the name of my request is quite long, I can adjust the width of the sidebar to fit the name. But the tabs don't show the full request names.
Another thing I find weird about tabs is that when I click on the directory in the sidebar, that directory also opens as a tab, even though it's completely unnecessary.

I would like to have an option to disable tabs to reduce visual clutter and make the interface cleaner.

**Comment by neopryn** [comment]
I have the same feedback as above.  I find the tabs a bit of an odd addition to the user interface when the request list is itself enough of an index for me to sort through requests, I find adds a lot of clutter personally.

A bit of feedback; I think the tabs would be useful if they specifically were for opening ephemeral historical requests without changing the base request template.  Frequently I find myself scrolling back through historical requests during API development (especially to test broken ones or find broken examples) but I like to keep requests in a semi-working template state, I don't know how they would contribute back to the history and its a very rough idea but I don't really see the point of tabs otherwise except for "let me quickly try this without ruining my collection".

**Comment by plopezlpz** [comment]
Is there an option to disable tabs? I used to like insomnia better than postman because of no tabs!

**Comment by morozover** [comment]
I need «Turn off tabs» option

**Reply by gopher-work** [reply]
I cannot agree more.. it's one of the reasons why I don't like postman and prefer Insomnia... is there a way to disable them, please

**Comment by chayglass84** [comment]
Crossposting my update from last week - 12.4 (GA last week) includes much, much better tabs

https://github.com/Kong/insomnia/issues/8488#issuecomment-3974285971
